### ISSUE TITLE
Multiple emit targets don’t ignore `-o` option, contrary to the warning

### ISSUE THREAD

**Issue by nagisa** [rust-lang/rust#20130]
```
$ ls -alh
…
drwxr-xr-x 1 nagisa nagisa  26 Dec 22 11:32 src/
…
$ rustc src/lib.rs --emit=llvm-ir,llvm-bc,obj,asm,link --crate-type=rlib -o rdd
warning: ignoring specified output filename because multiple outputs were requested
$ ls -alh
total 136K
…
-rw-r--r-- 1 nagisa nagisa  86K Dec 22 16:41 liblib.rlib
-rw-r--r-- 1 nagisa nagisa 5.1K Dec 22 16:41 rdd.bc
-rw-r--r-- 1 nagisa nagisa  17K Dec 22 16:41 rdd.ll
-rw-r--r-- 1 nagisa nagisa 9.7K Dec 22 16:41 rdd.o
-rw-r--r-- 1 nagisa nagisa  16K Dec 22 16:41 rdd.s
drwxr-xr-x 1 nagisa nagisa  26 Dec 22 11:32 src/
…
```

Although compiler warns that output filename was ignored, only the `link` target ignores it. All other targets generate file with provided output filename.

**Comment by thorncp** [comment]
6e7968b10a32cbef810dcb9e3d5bc736dbf61d1f suggests that this behavior is intended. Specifically:

> The -o option has also been redefined to be used for *all* flavors of outputs.
> This means that we no longer ignore it for libraries.

I think the warning message is misleading, and is probably not even necessary. What do we think about removing this warning entirely, and updating `rustc -h` and the man page to reflect this behavior?

**Comment by nagisa** [comment]
That commit message contradicts itself further down:

> If the -o flag is specified, and only one output type is specified, the
   output will be emitted at this location. If more than one output type is
   specified, then the filename of -o is ignored, and all output goes in the
   **directory** that -o specifies.

What’s the expected behaviour here @alexcrichton?

**Comment by brson** [comment]
I don't recall what was intended but it does seem that the current behavior is inconsistent and the warning message is wrong.

Based on the current behavior my preference is that if there are multiple output types that we name them all based on the provided file stem, meaning just change the library output to conform to the existing behavior.

@alexcrichton What do you think?

**Comment by nagisa** [comment]
This is possibly a good time and place to brainstorm about this, since this is one of the cornerstones of compiler usage.

This is what sounds most sensible to me:

* `--out-dir` *never* ignored;
* `-o` is alias for `--crate-name`.

```
$ # This is what it would look like
$ rustc --crate-type=rlib --out-dir=target/ -o foobar test.rs --emit=[everything] && ls target/
foobar.o foobar.s foobar.ll foobar.bc libfoobar.rlib
```

It does not depend on any combination of flags and works predictably (though very incompatible to other compilers) well.

**Comment by alexcrichton** [comment]
My thinking is the same as @brson's, the intended operation is something like:

```
// If foo.rs only generates one file, it will be called precisely `foo`
$ rustc foo.rs -o foo
// Use `foo` as a filestem for all outputs, ignore the precise filename "foo"
$ rustc --emit=[lots] foo.rs -o foo
```

The `--out-dir` options always takes precedence and will ignore `-o` entirely if specified, I believe. The major purpose of the `-o` flag is to compile executables such as `rustc foo.rs -o bar`.

**Comment by nagisa** [comment]
I submitted a https://github.com/rust-lang/rfcs/pull/595 regarding this.

**Comment by wthrowe** [comment]
Since we can't actually change this behavior anymore since it would be breaking, would a pull request that just gets rid of the incorrect warning be accepted?

**Comment by brson** [comment]
Yes, let's change this so that, only if `link` is one of the requested, it says: "warning: ignoring output name requested with -o for "link" output because multiple outputs were requested".

In Rust 2 we should consider changing it so all output types are affected by the rename.

**Comment by Manishearth** [comment]
Willing to help anyone who wants to work on this

**Comment by hgallagher1993** [comment]
Ya I'd take it

**Comment by brson** [comment]
@hgallagher1993 You got it!

**Comment by brson** [comment]
@hgallagher1993 How's it going?

**Comment by hgallagher1993** [comment]
@brson I actually haven’t had a chance to even look at it, long hours in the office all last week. It's going to have my undivided attention for the next few days :smiley: I'm probably going to need a bit of this as well @Manishearth

**Comment by brson** [comment]
OK, thanks for the status update!

**Comment by hgallagher1993** [comment]
@Manishearth I've been looking at this task and I was just wondering could I get some direction on how to start this...still completely new to the compiler :smile:

**Comment by Manishearth** [comment]
So the warning is here: https://github.com/rust-lang/rust/blob/6cc49e51de7ea9b0cc4aff437975544233c57107/src/librustc_driver/driver.rs#L1275

RIght now we warn if the number of unnamed output types is more than one. The actual condition is "if the number of output types is more than one AND if one of the unnamed emit options is `link`"

The emit options are stored in a hashmap in `sess.opts.output_types`. The current code just looks for hashmap entries that exist but have a `None` value. What we need to look for is an entry where the key is [`Exe`](http://manishearth.github.io/rust-internals-docs/rustc/session/config/enum.OutputType.html) and the value is `None`.

Something like `&& sess.opts.output_types.get(Exe).is_none` in that if should fix it. You will need to update the text to that from https://github.com/rust-lang/rust/issues/20130#issuecomment-228900232 too.

**Comment by kjaleshire** [comment]
I believe this issue is resolved. See #38840 .
